diff/windows/windows.go

```diff
@@ -332,6 +335,22 @@ func mountsToLayerAndParents(mounts []mount.Mount) (string, []string, error) {
 		return "", nil, err
 	}
 
+	if mnt.ReadOnly() {
+		if mnt.Type == "bind" && len(parentLayerPaths) != 0 {
+			return "", nil, fmt.Errorf("unexpected bind-mount View with parents: %w", errdefs.ErrInvalidArgument)
+		} else if mnt.Type == "bind" {
+			// rootfs.CreateDiff creates a new, empty View to diff against,
+			// when diffing something with no parent.
+			// This makes perfect sense for a walking Diff, but for WCOW,
+			// we have to recognise this as "diff against nothing"
+			return "", nil, nil
+		} else if len(parentLayerPaths) == 0 {
+			return "", nil, fmt.Errorf("unexpected windows-layer View with no parent: %w", errdefs.ErrInvalidArgument)
+		}
+		// Ignore the dummy sandbox.
```

Comment:
I just noticed that this comment's [partner in `createSnapshot`](https://github.com/containerd/containerd/pull/4419/files#diff-b6ae335f862c747466f1bc00e89eb1ef7cd9589bcc89ec57953c9c20125b3752R425-R427) has been lost as part of the refactoring to use BindFilter.

We also no-longer ignore the rootfs-size labels in the read-only case, although I hope no one passes such labels into View creation...

It might be worth adding a comment [at that point in the snapshotter](https://github.com/containerd/containerd/pull/8043/files#diff-b6ae335f862c747466f1bc00e89eb1ef7cd9589bcc89ec57953c9c20125b3752R420) that when generating mounts for Views, we still create a scratch layer, because we can't actually mount a stack of layers on Windows unless the top one is a scratch layer, and noting that the differ knows to ignore the scratch layer when comparing layers in such mounts. This comment could then directly refer to that comment, since if you're not reading the PR, it's not clear what "Ignore the dummy sandbox" is referring to.

To be clear, I am _assuming_ BindFilter still requires a top-most scratch layer to mount. In theory a locally-created layer is both scratch and read-only (we don't delete sandbox.vhdx as the hcs layer-differ consumes that from the top layer, not the exported format) so such a layer could perhaps be mounted directly as a view using BindFilter's read-only flag, and no dummy sandbox. We'd need to communicate that through to the differ somehow (or make the mounter more complex and handle it there), so consider such an exploration as a possible future optimisation that's not even strongly justified. (And merely mounting the sandbox may mutate it, which is pretty non-ideal.)

Reply:
Huh, for some reason I had the idea it could actually be given the list of layers to mount, and take care of WCIFS activation internally. Sorry, ignore that last paragraph then. ^_^